### PR TITLE
HIVE-28059: Iceberg REST Catalog

### PR DESCRIPTION
**Pull request by henrib**
### What changes were proposed in this pull request?
Expose a servlet running in a Jetty hosted by HMS that expose the Iceberg REST API

### Why are the changes needed?
To allow query engines to consume Iceberg tables without having to talk Thrift

### Does this PR introduce _any_ user-facing change?
Yes; the Iceberg API is documented

### Is the change a dependency upgrade?
No

### How was this patch tested?
Unit tests

### CODE REVIEW

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  private static Reference<Catalog> catalogRef;
+  static Catalog getLastCatalog() {
+    return catalogRef != null? catalogRef.get() :  null;
+  }
+
+  private HMSCatalogServer() {
+    // nothing
+  }
+
+  public static HttpServlet createServlet(SecureServletCaller security, Catalog catalog) throws IOException {
```
Comment (aturoczy) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:55]: Why are these servers are static? Maybe it has some meaning in Java, but in .NET it was an anti-pattern in server-side world.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+   * Convenience method to start a http server that only serves this servlet.
+   * @param conf the configuration
+   * @return the server instance
+   * @throws Exception if servlet initialization fails
+   */
+
+  public static Server startServer(Configuration conf, HMSCatalog catalog) throws Exception {
+    int port = MetastoreConf.getIntVar(conf, MetastoreConf.ConfVars.CATALOG_SERVLET_PORT);
+    if (port < 0) {
+      return null;
```
Comment (aturoczy) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:116]: It should be an exception imho.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.iceberg.rest;
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class LocationUtil {
+  private LocationUtil() {
+  }
+
+  public static String stripTrailingSlash(String path) {
+    Preconditions.checkArgument(path != null && path.length() > 0, "path must not be null or empty");
```
Comment (aturoczy) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java:29]: Big + for Preconditions check. I only see few times when somebody check the parameter accordingly

File: hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    String result;
+    for(result = path; result.endsWith("/"); result = result.substring(0, result.length() - 1)) {
```
Comment (aturoczy) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java:32]: IMHO it is too compact to read it.

File: hms-catalog/src/test/resources/auth/jwt/jwt-authorized-key.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (aturoczy) [inline hms-catalog/src/test/resources/auth/jwt/jwt-authorized-key.json:1]: Why does this key file belong to the repository? It seems like sensitive keys. Or is this needed for the testing?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 6]
```diff
@@ -732,10 +756,25 @@ public static void startMetaStore(int port, HadoopThriftAuthBridge bridge,
...
 
+  static Server startIcebergCatalog(Configuration configuration) {
+    try {
+      Class<?> iceClazz = Class.forName("org.apache.iceberg.rest.HMSCatalogServer");
+      Method iceStart = iceClazz.getMethod("startServer", Configuration.class);
+      return (Server) iceStart.invoke(null, configuration);
+    } catch (ClassNotFoundException xnf) {
+      return null;
+    } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+      LOG.error("error starting Iceberg REST server {}", e);
```
Comment (aturoczy) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:773]: Error: Unable to start Iceberg REST Catalog server. 
(Iceberg not the best in naming)

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PropertyServlet.java [hunk 7]
```diff
@@ -359,14 +372,34 @@ public static Server startServer(Configuration conf) throws Exception {
...
+    ServletContextHandler context = new ServletContextHandler(ServletContextHandler.NO_SESSIONS);
+    context.setContextPath("/");
+    ServletHolder servletHolder = new ServletHolder(servlet);
+    servletHolder.setInitParameter("javax.ws.rs.Application", "ServiceListPublic");
+    context.addServlet(servletHolder, "/*");
+    context.setVirtualHosts(null);
+    context.setGzipHandler(new GzipHandler());
+
+    Server httpServer =
+        new Server(PropertyUtil.propertyAsInt(System.getenv(), "REST_PORT", 8181));
```
Comment (aturoczy) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PropertyServlet.java:398]: 8181 is a pretty generic port. It could cause port conflict.

**Review by aturoczy — CHANGES_REQUESTED** [review]
First round. Currently some minor change request and questions

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java (hunk 2 above)
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:116]: No, this is by design, if you don't specify anything in the configuration, the server does not start. If you put a 0, the system will find a free port; if you use > 0, it will attempt starting on this port.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java (hunk 1 above)
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:55]: Creating the servlet is static because we pass parameters; I'm not sure what wrong you see. And about the server being a singleton, this is pretty common in Java. Metastore uses something equivalent.

File: hms-catalog/src/test/resources/auth/jwt/jwt-authorized-key.json (hunk 5 above)
Comment (henrib) [reply hms-catalog/src/test/resources/auth/jwt/jwt-authorized-key.json:1]: It is a resource for tests when we mock JWT flow.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PropertyServlet.java (hunk 7 above)
Comment (henrib) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/PropertyServlet.java:398]: That main() method is entirely in comments it seems.

**Comment by zhangbutao** [comment]
Thanks for the good feat! Haven't looked at the code yet. I know we can provide a rest server catalog with this PR now. Can we HS2 use rest catalog client to query iceberg tables provided by other 3rd rest server catalog?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java (hunk 6 above)
Comment (henrib) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java]: Updated code accordingly.

**Comment by henrib** [comment]
> Thanks for the good feat! Haven't looked at the code yet. I know we can provide a rest server catalog with this PR now. Can we HS2 use rest catalog client to query iceberg tables provided by other 3rd rest server catalog?

No, this is just HMS exposing Iceberg tables as REST catalogs.

File: common/pom.xml [hunk 8]
```diff
@@ -314,6 +314,10 @@
       <groupId>javolution</groupId>
       <artifactId>javolution</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-storage-api</artifactId>
```
Comment (deniskuzZ) [inline common/pom.xml:319]: why was this added to common?

File: hms-catalog/pom.xml [hunk 9]
```diff
@@ -0,0 +1,264 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (deniskuzZ) [inline hms-catalog/pom.xml:1]: why do we create a root-level module? can we move this under
`standalone-metastore/rest-catalog`

Comment (deniskuzZ) [reply hms-catalog/pom.xml:1]: in Hive for Iceberg tables we support HiveCatalog & HadoopCatalog, for external and ACID tables we have. different interface.

File: hms-catalog/pom.xml [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <log4j2.debug>false</log4j2.debug>
+    <hive.version>${project.parent.version}</hive.version>
+    <revision>${hive.version}</revision>
+  </properties>
+  <dependencies>
+
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-standalone-metastore-server</artifactId>
```
Comment (deniskuzZ) [inline hms-catalog/pom.xml:28]: why do you need metastore-server?

File: hms-catalog/pom.xml [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+      <version>${revision}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.iceberg</groupId>
+      <artifactId>iceberg-bundled-guava</artifactId>
+      <version>${iceberg.version}</version>
+    </dependency>
+
+    <!-- Intellij dislikes shaded Iceberg: uncomment the following block helps when coding/debugging -->
+<!--
```
Comment (deniskuzZ) [inline hms-catalog/pom.xml:63]: please cleanup the code

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java [hunk 12]
```diff
@@ -0,0 +1,605 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.iceberg.BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.TABLE_TYPE_PROP;
+
+
+public class HMSCatalog extends BaseMetastoreCatalog implements SupportsNamespaces, Configurable {
```
Comment (deniskuzZ) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java:79]: why do we need this? can't we use HiveCatalog for the same?

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+
+/**
+ * Original @ https://github.com/apache/iceberg/blob/main/core/src/test/java/org/apache/iceberg/rest/RESTCatalogServlet.java
+ */
+
+/**
+ * The RESTCatalogServlet provides a servlet implementation used in combination with a
+ * RESTCatalogAdaptor to proxy the REST Spec to any Catalog implementation.
+ */
+public class HMSCatalogServlet extends HttpServlet {
```
Comment (deniskuzZ) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java:57]: I think, nowadays no one uses Java Servlets, there is a much more modern stack for that: spring-boot

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSTableOperations.java [hunk 14]
```diff
@@ -0,0 +1,597 @@
...
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import static org.apache.iceberg.TableProperties.GC_ENABLED;
+
+/**
+ * TODO we should be able to extract some more commonalities to BaseMetastoreTableOperations to
+ * avoid code duplication between this class and Metacat Tables.
+ */
+public class HMSTableOperations extends BaseMetastoreTableOperations {
```
Comment (deniskuzZ) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/HMSTableOperations.java:85]: is that a clone of HiveTableOperations?

File: hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.rest;
+
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+public class LocationUtil {
```
Comment (deniskuzZ) [inline hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java:24]: is that copy-pasted from iceberg-core?

File: hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+ * Placed here in a separate class so it can be shared across unit tests.
+ */
+public final class TestTxnDbUtil {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TestTxnDbUtil.class.getName());
+  private static final String TXN_MANAGER = "org.apache.hadoop.hive.ql.lockmgr.DbTxnManager";
+
+  private static int deadlockCnt = 0;
+
+  private TestTxnDbUtil() {
```
Comment (deniskuzZ) [inline hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java:61]: it looks as a complete copy-paste of `org.apache.hadoop.hive.metastore.utils.TestTxnDbUtil`

File: iceberg/iceberg-catalog/pom.xml [hunk 17]
```diff
@@ -19,6 +19,23 @@
       <groupId>org.apache.hive</groupId>
       <artifactId>hive-iceberg-shading</artifactId>
     </dependency>
+
+    <!-- Intellij dislikes shaded Iceberg: uncomment the following block helps when coding/debugging -->
```
Comment (deniskuzZ) [inline iceberg/iceberg-catalog/pom.xml:23]: cleanup the code

File: hms-catalog/src/test/resources/scripts/metastore/upgrade/derby/hive-schema-4.0.0-beta-1.derby.sql [hunk 18]
```diff
@@ -0,0 +1,896 @@
+-- Timestamp: 2011-09-22 15:32:02.024
```
Comment (deniskuzZ) [inline hms-catalog/src/test/resources/scripts/metastore/upgrade/derby/hive-schema-4.0.0-beta-1.derby.sql:1]: why did you copied schema scripts here, who is gonna support them

File: iceberg/iceberg-shading/pom.xml [hunk 19]
```diff
@@ -37,8 +38,8 @@
     <dependency>
       <groupId>org.apache.hive</groupId>
       <artifactId>patched-iceberg-core</artifactId>
-      <optional>true</optional>
-    </dependency>
+      <version>patched-${iceberg.version}-${project.parent.version}</version>
```
Comment (deniskuzZ) [inline iceberg/iceberg-shading/pom.xml:41]: why is this required?

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/Optimizer.java [hunk 20]
```diff
@@ -192,7 +192,9 @@ public void initialize(HiveConf hiveConf) {
       transformations.add(new FixedBucketPruningOptimizer(compatMode));
     }
 
-    transformations.add(new BucketVersionPopulator());
+    if (!isTezExecEngine) {
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/Optimizer.java:195]: how is that related to the current patch?

File: serde/pom.xml [hunk 21]
```diff
@@ -249,6 +254,12 @@
       <version>${mockito-core.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-classification</artifactId>
+      <version>4.0.0-beta-2-SNAPSHOT</version>
```
Comment (deniskuzZ) [inline serde/pom.xml:260]: copy-paste

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java [hunk 22]
```diff
@@ -1808,6 +1808,27 @@ public enum ConfVars {
         "hive.metastore.properties.servlet.auth", "jwt",
         "Property-maps servlet authentication method (simple or jwt)."
     ),
+    PROPERTIES_NS("hive.metastore.properties.namespaces",
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:1811]: why do you put rest-catalog properties inside of HMS? shouldn't it have it's own configs?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java [hunk 23]
```diff
@@ -732,10 +756,26 @@ public static void startMetaStore(int port, HadoopThriftAuthBridge bridge,
     }
     // optionally create and start the property server and servlet
     propertyServer = PropertyServlet.startServer(conf);
+    // optionally create and start the Iceberg REST server and servlet
+    icebergServer = startIcebergCatalog(conf);
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java:760]: do we embed IcebergCatalog in HMS, what if I want to scale it differently?

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java (hunk 13 above)
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java:57]: Spring-boot is a huge dependency and the HMS-service, in its initial form, runs within HMS; embedding Jetty and implementing a servlet (copied from Iceberg) seem a simpler option.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogAdapter.java [hunk 24]
```diff
@@ -0,0 +1,588 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+/**
+ * Original @ https://github.com/apache/iceberg/blob/main/core/src/test/java/org/apache/iceberg/rest/RESTCatalogAdapter.java
+ */
+
+/** Adaptor class to translate REST requests into {@link Catalog} API calls. */
+public class HMSCatalogAdapter implements RESTClient {
```
[the remark this answers is not in the document]
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogAdapter.java:78]: Not strictly since we want the metrics and avoid shading. But it could move from test to src in the iceberg package I guess.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java (hunk 12 above)
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java:79]: HMSCatalog runs in-process within an Hive Metastore process/instance. The goal is an easy deployment as part of HMS. As you noticed, the HiveCatalog is also able to run the tests (adapted from Iceberg packages); in the future, it will be possible to run Hms-catalog as a separate process (being a thrift client).

File: hms-catalog/pom.xml (hunk 11 above)
Comment (henrib) [reply hms-catalog/pom.xml:63]: Could we remove the shading from the iceberg modules ?

File: hms-catalog/pom.xml (hunk 9 above)
Comment (henrib) [reply hms-catalog/pom.xml:1]: We are adding an Iceberg Catalog REST service to Hive Metastore so 3rd party query engines can rely on HMS to access the Iceberg metadata. (https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml)

File: common/pom.xml (hunk 8 above)
Comment (henrib) [reply common/pom.xml:319]: Probably a mistake, I'll investigate and correct.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java (hunk 22 above)
Comment (henrib) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:1811]: Because HMS starts the HMS-Catalog service.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/Optimizer.java (hunk 20 above)
Comment (henrib) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/Optimizer.java:195]: Mistake

File: iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 25]
```diff
@@ -489,18 +488,26 @@ protected String defaultWarehouseLocation(TableIdentifier tableIdentifier) {
   }
 
   private String databaseLocation(String databaseName) {
-    String warehouseLocation = conf.get(HiveConf.ConfVars.METASTORE_WAREHOUSE.varname);
-    Preconditions.checkNotNull(
-            warehouseLocation, "Warehouse location is not set: hive.metastore.warehouse.dir=null");
+    String warehouseLocation = conf.get("metastore.warehouse.dir");
```
[the remark this answers is not in the document]
Comment (henrib) [reply iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:491]: Yes, unfortunate older code (3.x) workaround.

File: iceberg/iceberg-catalog/pom.xml (hunk 17 above)
Comment (henrib) [reply iceberg/iceberg-catalog/pom.xml:23]: Took me a while to find why: 
https://issues.apache.org/jira/browse/HIVE-25027

File: hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java (hunk 16 above)
Comment (henrib) [reply hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java:61]: Yes

File: hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java (hunk 15 above)
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java:24]: Yes

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+    final String name;
+    String curi = configuration.get(MetastoreConf.ConfVars.THRIFT_URIS.getVarname());
+    String cwarehouse = configuration.get(MetastoreConf.ConfVars.WAREHOUSE.getVarname());
+    String cextwarehouse = configuration.get(MetastoreConf.ConfVars.WAREHOUSE_EXTERNAL.getVarname());
+    if ("HMSCatalog".equals(clazz)) {
+      name = "hms";
+      catalog = new HMSCatalog(configuration);
+    } else {
+      name = "hive";
+      catalog = new org.apache.iceberg.hive.HiveCatalog();
```
[the remark this answers is not in the document]
Comment (henrib) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:73]: HMSCatalog is using the RawStore to interact with metadata. (no thrift, in process, in memory).

File: hms-catalog/pom.xml (hunk 9 above)
Comment (deniskuzZ) [reply hms-catalog/pom.xml:1]: why can't we expose via REST existing HiveCatalog, why do we need to embed something into HMS?

File: hms-catalog/pom.xml (hunk 11 above)
Comment (deniskuzZ) [reply hms-catalog/pom.xml:63]: we can't, but that code is commented out

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java (hunk 12 above)
Comment (deniskuzZ) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java:79]: why should it run in HMS? idea was to have a REST interface that you can scale independently. Rest API should also support non-HMS catalogs: Hadoop, Jdbc, etc

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java (hunk 26 above)
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServer.java:73]: why? that is an independent component

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java (hunk 13 above)
Comment (deniskuzZ) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalogServlet.java:57]: in my opinion, this is poor design and lack of thought for future support.

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSTableOperations.java (hunk 14 above)
Comment (deniskuzZ) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSTableOperations.java:85]: 👎

File: hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java (hunk 15 above)
Comment (deniskuzZ) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/LocationUtil.java:24]: nice 👎

File: hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java (hunk 16 above)
Comment (deniskuzZ) [reply hms-catalog/src/test/java/org/apache/iceberg/rest/TestTxnDbUtil.java:61]: is that supposed to be a ready-for-review PR?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java (hunk 22 above)
Comment (deniskuzZ) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/conf/MetastoreConf.java:1811]: why in a world?

File: hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java (hunk 12 above)
Comment (SourabhBadhya) [reply hms-catalog/src/main/java/org/apache/iceberg/rest/HMSCatalog.java:79]: In the code, I see various application logic for retrieval (get list of namespaces, tables etc.) and manipulation (rename, alter etc.). This is done via HiveCatalog as well. Is there some required dependency  because of REST interface that we have to create a dedicated catalog which pretty much duplicates the logic already existing in an existing catalog?

**Comment by henrib** [comment]
Apologies for wasting your time.

**Comment by deniskuzZ** [comment]
> > Thanks for the good feat! Haven't looked at the code yet. I know we can provide a rest server catalog with this PR now. Can we HS2 use rest catalog client to query iceberg tables provided by other 3rd rest server catalog?
> 
> No, this is just HMS exposing Iceberg tables as REST catalogs.

some useful reading:
https://datastrato.ai/docs/0.3.1/iceberg-rest-service/
Works as a catalog proxy, supporting HiveCatalog and JDBCCatalog

**Comment by henrib** [comment]
It does not seem like Gravitino exposes the Iceberg REST API; they do expose their own. ( see https://github.com/datastrato/gravitino/tree/main/docs/open-api )
